test/proxy/fallback.js

```diff
@@ -0,0 +1,218 @@
+            },
+        );
+
+        assert.strictEqual(
+            await pmc.recordedBeneficiary.call(),
+            beneficiary,
+        );
+
+        assert.isOk(
+            (await pmc.recordedCurrencyCode.call()).eqn(currencyCode),
+        );
+
+        assert.isOk(
+            (await pmc.remainingBalance.call()).eqn(initialBalance - msgValue),
+        );
+
+        assert.strictEqual(
+            await mc.recordedBeneficiary.call(),
+            Utils.NULL_ADDRESS,
+        );
+
+        assert.isOk(
+            (await mc.recordedCurrencyCode.call()).eqn(0),
+        );
+
+        assert.isOk(
+            (await mc.remainingBalance.call()).eqn(initialBalance),
+        );
+    });
+
+    it('Checks that return data is correctly propagated back.', async () => {
+        const initialBalance = 22;
+
+        const mc = await MasterCopySpy.new(initialBalance);
+        const proxy = await ProxyContract.new(mc.address);
+        const pmc = await MasterCopySpy.at(proxy.address);
+        await pmc.setup(initialBalance);
+
+        const msgValue = 11;
+        const currencyCode = 7;
+        const beneficiary = accountProvider.get();
+
+        const remainingBalance = await pmc.pay.call(
+            currencyCode,
+            beneficiary,
+            {
+                value: msgValue,
+            },
+        );
+
+        assert.isOk(
+            remainingBalance.eqn(initialBalance - msgValue),
+        );
+    });
+
+    it('Checks that exception is handled properly.', async () => {
+        const initialBalance = 22;
+
+        const mc = await MasterCopySpy.new(initialBalance);
+        const proxy = await ProxyContract.new(mc.address);
+        const pmc = await MasterCopySpy.at(proxy.address);
+        await pmc.setup(initialBalance);
+
+        const currencyCode = 7;
+        const beneficiary = accountProvider.get();
+
+        await Utils.expectRevert(
+            pmc.pay.call(
+                currencyCode,
+                beneficiary,
+                {
+                    value: initialBalance + 1,
+                },
+            ),
+        );
+    });
+
+    it('Checks the event emition.', async () => {
+        const initialBalance = 22;
+
+        const mc = await MasterCopySpy.new(initialBalance);
+        const proxy = await ProxyContract.new(mc.address);
+        const pmc = await MasterCopySpy.at(proxy.address);
+        await pmc.setup(initialBalance);
+
+        const msgValue = 11;
+        const currencyCode = 7;
+        const beneficiary = accountProvider.get();
+
+        const transactionResponse = await pmc.pay(
+            currencyCode,
+            beneficiary,
+            {
+                value: msgValue,
+            },
+        );
+
+        const events = Event.decodeTransactionResponse(
+            transactionResponse,
+        );
+
+        assert.strictEqual(
+            events.length,
+            1,
+        );
+
+        Event.assertEqual(events[0], {
```

Comment:
would be good to additionally assert in which contract the event is emitted.

Reply:
Agree. I am planing to do this for all other event testing. There is a ticket in my list. Would suggest to do this after `proxy`, `credit` and `delayed recovery module` functionalities are merged.